### PR TITLE
[Qwen] Handle tie_word_embeddings for Qwen MoE and Qwen3Next

### LINKED ISSUE #19908: [Bug] Using SGlang for inference always outputs token id 0
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

I used slime for MoE SFT, and the loss was normal during the training process. 
When convert the Megatron checkpoint to hf checkpoint and conducting inference verification using SGlang at the end of the training, it was found that the model was unable to output normally. 
The specific manifestation is that regardless of the input, only the token "index:0" is output. Corresponding to the Qwen vocabulary, it is "!". At the same time, the logp value is also approximately -11. (For this step, I also conducted a verification using Llamafactory. The stored checkpoints that do not involve Megatron conversion still have this issue). 
However, this problem does not occur when switching to vllm, and normal output can be achieved. 
Since we will continue to use the RL framework based on SGlang for subsequent training, which really troubles me. I hope sombody can take a look and help me out.

Furthermore, since my model is based on the non-standard structure of Qwen3MoE (which is not any of the types of Qwen available for open-source), and because the current work is still ongoing and it is not convenient to open-source the specific checkpoints, the config.json is as follows: 

```json
{
  "architectures": [
    "Qwen3MoeForCausalLM"
  ],
  "attention_bias": false,
  "attention_dropout": 0.0,
  "bos_token_id": 151643,
  "decoder_sparse_step": 1,
  "dtype": "float32",
  "eos_token_id": 151643,
  "head_dim": 128,
  "hidden_act": "silu",
  "hidden_size": 1024,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "max_position_embeddings": 32768,
  "max_window_layers": 28,
  "mlp_only_layers": [],
  "model_type": "qwen3_moe",
  "moe_intermediate_size": 384,
  "norm_topk_prob": true,
  "num_attention_heads": 16,
  "num_experts": 232,
  "num_experts_per_tok": 8,
  "num_hidden_layers": 28,
  "num_key_value_heads": 8,
  "output_router_logits": false,
  "qkv_bias": false,
  "rms_norm_eps": 1e-06,
  "rope_scaling": null,
  "rope_theta": 1000000,
  "router_aux_loss_coef": 0.001,
  "sliding_window": null,
  "tie_word_embeddings": true,
  "transformers_version": "4.57.1",
  "use_cache": true,
  "use_qk_norm": true,
  "use_sliding_window": false,
  "vocab_size": 151936
}

```

And I have verified that, using the config.json for random initialization and using sglang for inference, the above situation still occurs.

sglang inference result:
Image

vllm inference result:

Image

### Reproduction

CUDA_VISIBLE_DEVICES=0,1,2,3 python3 -m sglang.launch_server --model-path /model_path --host 0.0.0.0 --port 40003 --tp 4 --chunked-prefill-size 8192 --mem-fraction-static 0.6

CUDA_VISIBLE_DEVICES=4,5,6,7 vllm serve /model_path --host 0.0.0.0 --port 40002 --tensor-parallel-size 4 --max-num-batched-tokens 8192 --gpu-memory-utilization 0.6

### Environment

I used both SGlang 0.5.2 and 0.5.9(latest) and it still have problems. vllm version is 0.8.4.

### PR DESCRIPTION
**Pull request by xingsy97**
## Motivation

Fixes #19908

Qwen3MoE, Qwen2MoE, and Qwen3Next always create a separate `ParallelLMHead`, ignoring `config.tie_word_embeddings`. When `True`, the checkpoint has no `lm_head.weight`, so `lm_head` stays randomly initialized. Other Qwen models (qwen2, qwen3, qwen2_vl, qwen2_5_vl, qwen3_vl, etc.) already handle this correctly.

## Modifications

Follow the pattern in `qwen3.py`:

- **`qwen3_moe.py` / `qwen2_moe.py`**: Add PP-aware tie logic in `__init__` and embed→lm_head weight copy in `load_weights`.
- **`qwen3_next.py`**: Tie `lm_head` to `embed_tokens` in `__init__`, skip `lm_head.weight` in `load_weights`, guard `set_embed_and_head` against deleting the same object twice.

## Accuracy Tests

N/A, weight loading logic only.

## Benchmarking and Profiling

N/A, no change to inference hot path.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

### CODE REVIEW

File: python/sglang/srt/models/qwen3_moe.py [hunk 1]
```diff
@@ -914,13 +914,20 @@ def __init__(
...
+            else:
+                self.lm_head = ParallelLMHead(
+                    config.vocab_size,
+                    config.hidden_size,
+                    quant_config=quant_config,
+                    prefix=add_prefix("lm_head", prefix),
+                    use_attn_tp_group=get_global_server_args().enable_dp_lm_head,
+                )
+        else:
+            self.lm_head = PPMissingLayer()
```
[the remark this answers is not in the document]
Comment (xingsy97) [reply python/sglang/srt/models/qwen3_moe.py:963]: they're independent

File: python/sglang/srt/models/qwen3_moe.py [hunk 2]
```diff
@@ -1048,6 +1055,16 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
...
+            # PP tied: copy embed_tokens into lm_head, then fall through.
+            if name == "model.embed_tokens.weight":
+                if self.pp_group.is_last_rank and self.config.tie_word_embeddings:
+                    if "lm_head.weight" in params_dict:
+                        param = params_dict["lm_head.weight"]
+                        weight_loader = getattr(
+                            param, "weight_loader", default_weight_loader
+                        )
+                        weight_loader(param, loaded_weight)
+
```
[the remark this answers is not in the document]
Comment (xingsy97) [reply python/sglang/srt/models/qwen3_moe.py:1117]: they're independent
